Dispatcher result refactored, RouteCollector to RouteCollection renamed

burzum opened the pull request:
* The dispatcher returns now a result object (I don't mind discussing the method names)
  * ArrayAccess is provided for backward compatibility.
* The result object contains now always a route object, this is a preparation for more changes related to give this object more meaning and power.
* RouteCollector renamed to RouteCollection.
  * This would break BC if somebody uses the class directly but renaming it should no big deal at all. An alias could provided as well.
* Refactored data generators

lcobucci commented on .editorconfig:
```diff
@@ -0,0 +1,27 @@
+trim_trailing_whitespace = true
+charset = utf-8
+
+[*.bat]
+indent_style = space
+indent_size = 4
+end_of_line = crlf
```
We don't have `.bat` files here so this is unnecessary

lcobucci commented on .editorconfig:
```diff
@@ -0,0 +1,27 @@
+; This file is for unifying the coding style for different editors and IDEs.
```
This is very useful but I feel like we should introduce this in another PR.

burzum replied:
Why another PR? Whats the argument for that? Of course I can delete the file and add it back in another PR but its pointless and causes me just more work, just to get the same thing added back?

lcobucci replied:
Splitting things that are unrelated to this work makes everything easier for everyone. If you don't want to operate in a collaborative manner, than you shouldn't expect maintainers do to the same.

lcobucci commented on .editorconfig:
```diff
@@ -0,0 +1,27 @@
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+charset = utf-8
```
```suggestion
charset = utf-8
indent_size = 4
```

Adding `indent_size` here removes the need for configuring it for different extensions.

lcobucci commented on src/Dispatcher.php:
```diff
@@ -11,14 +11,10 @@ interface Dispatcher
      * Dispatches against the provided HTTP method verb and URI.
-     *
      * Returns array with one of the following formats:
-     *
      *     [self::NOT_FOUND]
      *     [self::METHOD_NOT_ALLOWED, ['GET', 'OTHER_ALLOWED_METHODS']]
      *     [self::FOUND, $handler, ['varName' => 'value', ...]]
```
This is outdated now

lcobucci commented on src/IRoute.php:
```diff
@@ -0,0 +1,19 @@
+
+namespace FastRoute;
+
+interface IRoute
```
I'd avoid using prefixes too. This should be simply `Route` and the implementation can live in another namespace (and also use `Route` as name).

The main point is that the interface should be the connection point everywhere.

burzum replied:
I would prefer a suffix, the full word "Interface" just by looking the files its always a PITA to identify interfaces if there is no prefix or suffix. So I would vote vor changing it everywhere to add a prefix or suffix.

lcobucci replied:
That goes against what we have in the project. Nothing uses prefixes nor suffixes because we don't see value in differentiating these components in this way.

lcobucci commented on src/Result.php:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @implements ArrayAccess<int, mixed>
+ */
+class Result implements ArrayAccess
```
Do we really need to make this behave as an array? What's the added value?

lcobucci replied:
Also, what's the added value of enabling inheritance here? To me this is basically a transport object and could be `final`.

burzum replied:
It's a lib and people might alter it and want to change the object. I'm not a big fan of using final classes and private properties in libs but prefer to keep them open. My experience is that even if people think "nobody" ever needs it, there are cases in which people will change it for a good reason.

lcobucci replied:
This is essentially a DTO. It can be final. If we want people to extend this we should provide an interface. I'd prefer to start the process with just the object. 

This patch breaks BC by renaming the classes, it doesn't make sense not to simplify things with that argument.

lcobucci commented on src/Result.php:
```diff
@@ -0,0 +1,183 @@
+    public function handler()
+    {
+        if (! isset($this->result[1])) {
+            return null;
+        }
+
+        return $this->result[1];
+    }
+
+    public function status(): int
+    {
+        return $this->status;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function args()
+    {
+        if (! isset($this->result[2])) {
+            return [];
+        }
+
+        return $this->result[2];
+    }
+
+    public function routeMatched(): bool
+    {
+        return $this->result[0] === self::FOUND;
+    }
+
+    public function methodNotAllowed(): bool
+    {
+        return $this->result[0] === self::METHOD_NOT_ALLOWED;
+    }
+
+    public function routeNotFound(): bool
+    {
+        return $this->result[0] === self::NOT_FOUND;
+    }
```
I don't think we should use `$this->result` for these methods. In fact, that property shouldn't exist IMHO.

burzum replied:
Backward compatibility, see my other comment.

lcobucci commented on src/Result.php:
```diff
@@ -0,0 +1,183 @@
+     */
+    public function toArray(): array
+    {
+        return $this->result;
```
I believe we should create the array dynamically based on the object state.

burzum replied:
The array should already have only what is needed. Since the tests still pass I think this is true.

lcobucci replied:
Yeah, we're holding something in memory, which is only useful for this method. It doesn't make sense to me.

lcobucci commented on src/Result.php:
```diff
@@ -0,0 +1,183 @@
+    protected $matched = false;
+
+    /** @var IRoute */
+    protected $route;
```
This is a write-only property, why do we need it here?

burzum replied:
Because you can get the route object of the route that matched. I'm just waiting what happens to this PR to see if its worth to spend more of my time with this or not. If it's worth the time I'll continue and refactor the routes and make sure a route object is used everywhere.

lcobucci replied:
So let's break things down and provide a full implementation of something instead of half steps.

burzum replied:
The property already contains the route object, there is nothing to wait on.

lcobucci commented on src/Route.php:
```diff
@@ -5,12 +5,12 @@
     public $httpMethod;
 
-    /** @var string */
+    /** @var ?string */
```
I think this is incorrect (since the constructor disallow `null` here)

burzum replied:
Good catch

lcobucci commented on src/Route.php:
```diff
@@ -19,25 +19,61 @@ class Route
     /**
      * Tests whether this route matches the given string.
+     *
+     * @param string $string String
```
Not a very useful annotation here :)

lcobucci reviewed (commented):
@burzum sorry about my delay. I really like the direction we're going here, I'd just suggest to split things into different PRs (e.g.: `RouteCollector -> RouteCollection` makes sense to me and I'd merge it right away).

I don't have many instructions on how we want to progress with v2.0 so instead of waiting for @nikic too I'd follow my instincts and Nikita can send patches/issues to correct the path.

burzum commented:
@lcobucci so yet another month has passed, any chance I'm getting answers to my comments please? By this speed we'll have a new version sometimes around 2050. 😭

lcobucci commented:
> @lcobucci so yet another month has passed, any chance I'm getting answers to my comments please? By this speed we'll have a new version sometimes around 2050. 😭 

First of all, this is OSS. We're not on the clock to ship anything. I'm doing my best to help this go through. If that isn't enough for you, sorry, but that's what I have to offer for now.

burzum commented:
@lcobucci just because something is OSS doesn't mean that somebody who is an *active* maintainer jumps in every few month. A maintainer has responsibility as well. It is very demotivating for the people who contribute and won't receive a response within a reasonable timeframe. Especially if you invest time, that is not covered by a company, and then get more or less ignored for weeks. I would prefer a statement like "Hey, I don't care!" than having to wait and trying to make the changes in a way *everyone* gets something out of it. I could also just make them for myself and don't care aboute contributing back and give a 💩 as well. 🤷 

This PR is green, it should be almost fully BC. This could be released as 2.0 and then a 3.0 could be started to remove the deprecated things and make more breaking changes. This way it will provide a smooth migration path.

lcobucci commented:
@burzum I understand your complaints, expectations, and demotivation. The part that it's being frustrating and hard in this process is to help you understand my side as well.

What made me bring the OSS argument is the fact that my time is just as important as yours. And I feel disrespected when, despite having asked you to split things up in different PRs so we can have more focused discussions, you kept increasing the scope. 

We don't know each other and we have very little context about what happens behind the keyboards on each side of this PR. That also doesn't help much 😅

Our approaches to development seem to be quite different but we also seem to care about similar things.

My general opinion about this PR didn't change. I feel like it has too many things. There are great ideas too. I just disagree with execution and the impact of that on the design of the lib - hence my requests to split and work together to find a good balance.

I'd love to have a synchronous chat with you to make things move forward and (hopefully) be able to share with you a little about my side and show you that I do care.

burzum commented:
@lcobucci OK... Any suggestion on how to split this set of changes up without causing me a ton of more work? I don't want to spend hours just to break it up.

lcobucci commented:
@burzum I'll have a closer look at it tomorrow and suggest something - I can definitely help with git stuff to make sure things don't take long.

lcobucci commented:
My proposal for splitting this is to incrementally extract the changes related to each topic.

We'd start by squashing all the changes in a single commit on a reference branch, then we create a branch for the first topic we'll extract (based on the reference branch). The extraction consists of resetting that commit, picking the necessary changes, logically committing things, and sending a PR.

Once the PR is created, the reference branch is rebased to remove the extracted changes from the reference commit.
After merging the PR, the reference branch is synced again just to make sure we're good.

---

We can share the load on extracting the code - I've done this kind of thing several times, so am quite comfortable with that. We can also try to do that together via any tool (google meet/zoom/`${your-tool-of-choice-here}`). I'd like for you to have your name on the stuff you're doing, so I'd prefer to have PRs created by you from your fork.

If you agree, we can convert this PR into a draft to make it the reference and start the extraction. I'd start with the `.editorconfig` and the using an object as the dispatching result.

What do you think?

burzum commented:
I won't have much time this week, next week maybe? Thuesday evening? Zoom, Skype, Slack, Google, whatever you want, after more than 12 month pandemic I have them all. 😆

By the way, meanwhile I've [got an ugly but working prototype of reverse routing](https://github.com/burzum/FastRoute/blob/fastroute-ng/src/Route.php#L80-L123). But I've ditched most BC except the BC of the result of the dispatcher in that branch and also replaced the creation of the result object with a factory and renamed the interfaces to suffix them with interface, so they match the standards I'm used to. I dislike when they're discarded, makes it easy to think it's a class, even with an IDE using shiny icons.

https://github.com/burzum/FastRoute/blob/fastroute-ng/src/

lcobucci commented:
Next week sounds good to me :+1: 
Ping on me [Twitter](https://twitter.com/lcobucci) or Discord (lcobucci#1456).

In the meantime, I'll rebase this branch to have it squashed and synced to `master` :+1:

lcobucci commented:
The new CI (GH-based) is now failing due to some new and advanced checks I added.
We'll solve them during the extraction process, so let's not worry too much now :+1:

burzum commented:
@lcobucci do you want the changes from https://github.com/burzum/FastRoute/blob/fastroute-ng/src/ as well? I've added reverse routing and named routes and finished getting everything into factories. Architecture wise it's done, I might just need to refine some methods and find edge cases. Feature wise the refactor is completed. I've added there typed properties, so it won't work below 7.4. 7.3 is EOL by the way.  It contains everything I wanted for now.

burzum commented:
I fixed all phpstan errors and some doc blocks in https://github.com/burzum/FastRoute/blob/fastroute-ng/src/
